src/clients/c++/simple_ensemble_client.cc

```diff
@@ -0,0 +1,294 @@
+    const std::vector<std::string>& filenames, const size_t batch_size)
+{
+  if (results.size() != 1) {
+    std::cerr << "expected 1 result, got " << results.size() << std::endl;
+    exit(1);
+  }
+
+  const std::unique_ptr<nic::InferContext::Result>& result =
+      results.begin()->second;
+
+  for (size_t b = 0; b < batch_size; ++b) {
+    size_t cnt = 0;
+    nic::Error err = result->GetClassCount(b, &cnt);
+    if (!err.IsOk()) {
+      std::cerr << "failed reading class count for batch " << b << ": " << err
+                << std::endl;
+      exit(1);
+    }
+
+    std::cout << "Image '" << filenames[b] << "':" << std::endl;
+
+    for (size_t c = 0; c < cnt; ++c) {
+      nic::InferContext::Result::ClassResult cls;
+      nic::Error err = result->GetClassAtCursor(b, &cls);
+      if (!err.IsOk()) {
+        std::cerr << "failed reading class for batch " << b << ": " << err
+                  << std::endl;
+        exit(1);
+      }
+
+      std::cout << "    " << cls.idx << " (" << cls.label << ") = " << cls.value
+                << std::endl;
+    }
+  }
+}
+
+void
+Usage(char** argv, const std::string& msg = std::string())
+{
+  if (!msg.empty()) {
+    std::cerr << "error: " << msg << std::endl;
+  }
+
+  std::cerr << "Usage: " << argv[0] << " [options] <image filename / image folder>" << std::endl;
+  std::cerr << "\t-v" << std::endl;
+  std::cerr << "\t-i <Protocol used to communicate with inference service>"
+            << std::endl;
+  std::cerr << "\t-u <URL for inference service>" << std::endl;
+  std::cerr << "\t-p <proprocessed output filename prefix>" << std::endl;
+  std::cerr << std::endl;
+  std::cerr
+      << "For -i, available protocols are 'grpc' and 'http'. Default is 'http."
+      << std::endl;
+  std::cerr
+      << "For -p, if specified, the client will only send the raw images to the"
+      << " preprocess custom backend, and write the response to files as output"
+      << " in <prfix><batch index> format."
+      << std::endl;
+
+  exit(1);
+}
+
+}  // namespace
+
+int
+main(int argc, char** argv)
+{
+  bool verbose = false;
+  std::string url("localhost:8000");
+  std::string protocol = "http";
+  std::string prefix = "";
+
+  // Parse commandline...
+  int opt;
+  while ((opt = getopt(argc, argv, "vi:u:p:")) != -1) {
+    switch (opt) {
+      case 'v':
+        verbose = true;
+        break;
+      case 'i':
+        protocol = optarg;
+        break;
+      case 'u':
+        url = optarg;
+        break;
+      case 'p':
+        prefix = optarg;
+        break;
+      case '?':
+        Usage(argv);
+        break;
+    }
+  }
+
+  nic::Error err;
+
+  // We use a simple model that takes 2 input tensors of 16 strings
+  // each and returns 2 output tensors of 16 strings each. The input
+  // strings must represent integers. One output tensor is the
+  // element-wise sum of the inputs and one output is the element-wise
+  // difference.
+  std::string model_name = "image_preprocess";
```

Comment:
This should use the ensemble model that includes image preprocessing followed by RN50 (or some other image classification model). This test will have to go on gitlab since it uses a RN50 model and those aren't available on github. We have a problem in that for someone to run this as an example they will have to create a model store with the custom preprocess model, the ensemble model, and a RN50 model. We should talk about the best way to add this CI test.